hera_cal/tests/test_nucal.py

```diff
@@ -561,3 +558,261 @@ def test_fit_nucal_foreground_model():
+
+    def test_compute_filters(self):
+        assert self.frc._filters_computed == False
+
+        # Compute filters
+        self.frc._compute_filters(self.freqs, 20e-9)
+        assert self.frc._filters_computed == True
+
+        # Trigger check that filters are already computed
+        sf = deepcopy(self.frc.spectral_filters)
+        self.frc._compute_filters(self.freqs, 20e-9)
+        assert np.allclose(self.frc.spectral_filters, sf)
+
+        # Recompute with different values 
+        self.frc._compute_filters(self.freqs, 30e-9)
+        assert sf.shape != self.frc.spectral_filters.shape
+
+        # Check that filters are the correct shape
+        assert len(self.frc.spatial_filters) == len(self.radial_reds[0])
+
+    def test_estimate_degeneracies(self):
+        # Copy data
+        data = deepcopy(self.data)
+
+        # Compute filters
+        self.frc._compute_filters(self.freqs, 20e-9)
+
+        # Fit model
+        model = nucal.fit_nucal_foreground_model(
+            data, self.data_wgts, self.radial_reds, self.frc.spatial_filters, 
+            return_model_comps=False, share_fg_model=True
+        )
+
+        # Estimate degeneracies
+        amplitude, tip_tilt = self.frc._estimate_degeneracies(data, model, self.data_wgts)
+
+        # Since the data are already "perfectly calibrated" and the model should match the data to high precision, 
+        # the estimated tip-tilt should be zero and the amplitude should be 1
+        assert np.isclose(
+            np.sqrt(np.mean(np.square(amplitude['nn'] - 1))), 0, atol=1e-5
+        )
+        assert np.isclose(
+            np.sqrt(np.mean(np.square(tip_tilt['nn']))), 0, atol=1e-5
+        )
+    
+        # Shape of amplitude and tip-tilt should be (ndims, ntimes, nfreqs)
+        assert tip_tilt["nn"].shape == (1, 2, 200)
+
+    def test_post_redcal_nucal(self):
+        """
+        """
+        amp = np.random.normal(1, 0.05, size=(2, 200))
+        gains = {(k, "Jnn"): amp for k in self.antpos}
+        dc = deepcopy(self.data)
+        apply_cal.calibrate_in_place(dc, gains, gain_convention='multiply')
+
+        fit_gains, model_params, meta = self.frc.post_redcal_nucal(
+            dc, self.data_wgts, spatial_estimate_only=True, minor_cycle_maxiter=10,
+            share_fg_model=True, maxiter=10, spectral_filter_half_width=10e-9,
+        )
+
+        # Check that the value of the loss function decreases
+        assert meta['nn']['loss_history'][-1] < meta['nn']['loss_history'][0]
+
+        fit_gains, model_params, meta, model = self.frc.post_redcal_nucal(
+            dc, self.data_wgts, spatial_estimate_only=True, minor_cycle_maxiter=10,
+            share_fg_model=True, maxiter=10, return_model=True, 
+            spectral_filter_half_width=10e-9, estimate_degeneracies=True
+        )
+
+        # Check that the value of the loss function decreases
+        assert meta['nn']['loss_history'][-1] < meta['nn']['loss_history'][0]
```

Comment:
Seems like you can do better than this... Can't you check that nucal gets the right answer to within some margin of error?

Reply:
Fixed this